[aliciamichellew] iP

aliciamichellew opened the pull request:
### Kevin Chatbot

> Kevin frees your mind of having to remember things you need to do. 

It's,

- text-based
- _easy_ to learn
- **SUPER FAST** to use

All you need to do is,

1. open the [program](https://github.com/aliciamichellew/ip).
2. add your tasks. (use `todo`, `event` or `deadline` keywords)
3. let it manage your tasks for you 😉

And it is FREE!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

Some code preview:

```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aliciamichellew committed:
Change bot name and add greetings

aliciamichellew committed:
Abstract out horizontal lines

aliciamichellew committed:
Add echo functionality

aliciamichellew committed:
Add ToDoList functionality

aliciamichellew committed:
Add mark as done functionality

aliciamichellew committed:
Refactor ToDo to Task and ToDoList to TaskList

aliciamichellew committed:
Make ToDo, Event, Deadline class

aliciamichellew committed:
Add ToDo, Event, Deadline functionalities

aliciamichellew committed:
Add testing

aliciamichellew committed:
Intialise exceptions

aliciamichellew committed:
Initialise enums

aliciamichellew committed:
Refactor and add error catch

aliciamichellew committed:
Add delete functionality

aliciamichellew committed:
Initialise logger

aliciamichellew committed:
Remove unused functions

aliciamichellew committed:
Refactor

aliciamichellew committed:
Delete unused exceptions

aliciamichellew committed:
Clean up import lines and comments

aliciamichellew committed:
Refactor evaluator

aliciamichellew committed:
Initialise storage

aliciamichellew committed:
Add save file functionality

aliciamichellew committed:
Remove unused function

aliciamichellew committed:
Add change deadline to datetime

aliciamichellew committed:
Fix datetime bug

aliciamichellew committed:
Change Logger package to Ui package

aliciamichellew committed:
Move packages to kevin package

aliciamichellew committed:
Merge branch 'add-gradle-support'

aliciamichellew committed:
Add JUnit test case

aliciamichellew committed:
Modify build gradle

aliciamichellew committed:
Add JavaDoc for BaseStrategy

aliciamichellew committed:
Add JavaDoc for ByeStrategy

aliciamichellew committed:
Add JavaDoc for DeadlineStrategy

aliciamichellew committed:
Add JavaDoc for DeleteStrategy

aliciamichellew committed:
Add JavaDoc for Evaluator

aliciamichellew committed:
Add JavaDoc for EventStrategy

aliciamichellew committed:
Add JavaDoc for ListStrategy

aliciamichellew committed:
Add JavaDoc for MarkStrategy

aliciamichellew committed:
Add JavaDoc for ToDoStrategy

aliciamichellew committed:
Add JavaDoc for UnmarkStrategy

aliciamichellew committed:
Add JavaDoc for KevinException

aliciamichellew committed:
Add JavaDoc for Commands

aliciamichellew committed:
Add JavaDoc for FileParser

aliciamichellew committed:
Add JavaDoc for Parser

aliciamichellew committed:
Add JavaDoc for QueryObject

aliciamichellew committed:
Add JavaDoc for FileStorage and Storage

aliciamichellew committed:
Add JavaDoc for Deadline

aliciamichellew committed:
Add JavaDoc for Event

aliciamichellew committed:
Add JavaDoc for Task

aliciamichellew committed:
Add JavaDoc for TaskList

aliciamichellew committed:
Add JavaDoc for ToDo

aliciamichellew committed:
Add JavaDoc for Logger

aliciamichellew committed:
Add JavaDoc for Kevin

aliciamichellew committed:
Change Commands to Command

aliciamichellew committed:
Add find functionality

aliciamichellew committed:
Merge branch 'branch-Level-9'

aliciamichellew committed:
Merge branch 'branch-A-CodingStandard'

aliciamichellew committed:
Change Commands to Command in Find

aliciamichellew committed:
Add JavaDoc for FindStrategy

gongg21 commented on src/main/java/kevin/Kevin.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class Kevin {
+    public static void main(String[] args) {
+        Logger logger = new Logger();
+        Parser parser = new Parser();
+        FileParser fileParser = new FileParser();
+        FileStorage fileStorage = new FileStorage();
+        TaskList taskList = new TaskList();
+        Scanner scanner = new Scanner(System.in);
+        Evaluator evaluator = new Evaluator(logger, fileStorage, taskList);
```
I like how you abstracted different types of objects and organised them. Any reason why you did it this way instead of minimising the number of classes?

gongg21 commented on src/main/java/kevin/parser/Parser.java:
```diff
@@ -0,0 +1,87 @@
+            }
+            args.add(querySplit[1]);
+        }
+        return new QueryObject(command, args);
```
Very detailed queries that assigns variables accurately and accordingly 👍

gongg21 commented on src/main/java/kevin/evaluator/Evaluator.java:
```diff
@@ -0,0 +1,65 @@
+    public boolean evaluate(QueryObject queryObject, boolean isInFile) throws KevinException {
+        Command command = queryObject.getCommandType();
+        ArrayList<String> arguments = queryObject.getArgs();
+        return MAPPER.get(command).apply(taskList, arguments, logger, fileStorage, isInFile);
```
Any reason why you chose to separate the different strategies to different classes rather than to specify actions under one class?

gongg21 reviewed (commented):
Looks well designed!
Just asked some questions for clarification.

phiphi-tan commented on src/main/java/kevin/Kevin.java:
```diff
@@ -0,0 +1,56 @@
+package kevin;
+
+import kevin.evaluator.Evaluator;
```
I like your very extensive abstractions into different relevant packages

phiphi-tan commented on src/main/java/kevin/Kevin.java:
```diff
@@ -0,0 +1,56 @@
+        }
+
+
+        while (true) {
```
Maybe you could consider whether all these actions be possibly passed into separate functions (e.g. command() / startup() ) to abstract their functions away from each other?

phiphi-tan commented on src/main/java/kevin/evaluator/BaseStrategy.java:
```diff
@@ -0,0 +1,38 @@
+     * @throws KevinException On the detection of errors.
+     */
+    public boolean evaluate(Logger logger, FileStorage fileStorage, boolean isInFile) throws KevinException {
+        return false;
```
I'm slightly confused with the return of this function; is this a work in progress or is it meant to always return false on the evaluate() call?

Noticed that the other classes implemented and overrid this method, would it be preferable to make this method abstract?

phiphi-tan commented on src/main/java/kevin/ui/Logger.java:
```diff
@@ -0,0 +1,42 @@
+
+    public static final String HORIZONTAL_LINE = "_".repeat(70);
+
+    public static final String BOT_NAME = "\t \n" +
```
really like the ASCII art

phiphi-tan reviewed (commented):
Well made and polished project with an extensive amount of abstraction into different packages. Just a few personal opinions / suggestions, cheers

aliciamichellew committed:
Add checkstyle config files

aliciamichellew committed:
Fix to pass checkstyle

aliciamichellew committed:
Setup JavaFX

aliciamichellew committed:
Make Gui until iteration 2

aliciamichellew committed:
Implement gui until iteration 3

aliciamichellew committed:
Implement fxml

aliciamichellew committed:
Integrate gui with command

aliciamichellew committed:
Fix a few UI

aliciamichellew committed:
GUI: make programme exit after bye command

Previously, only the GUI window will exit and the programme keeps running and needs to be manually killed.

Let's change this behaviour to kill programme after closing the window.

aliciamichellew committed:
File Parser: Fix parsing for event and deadline

Previously, event and deadline are saved in the text file. However, upon the launch of the programme,
the tasks are not added to the tasklist due to parsing error.

Let's fix this so that event and deadline can be parsed correctly and added to the taskList.

aliciamichellew committed:
File Storage: Fix deletion when there are duplicate tasks

Previously, there is a bug: if there are duplicate tasks, when deleting one task, it will result to other
duplicated tasks to be deleted as well.

Let's fix this so that it will only remove the task with that index.

aliciamichellew committed:
Fix deletion of task in file

aliciamichellew committed:
Fix parsing of event and deadline

aliciamichellew committed:
Handle mark and unmark when index out of bounds

aliciamichellew committed:
Format DialogBox file

aliciamichellew committed:
Fix delete index out of bound

Previously when deleting an index that is out of bound, it will not throw any error.

Let's fix this and throw an error if the user inserts an invalid index.

aliciamichellew committed:
Fix date parsing in deadline

aliciamichellew committed:
Add assertions

In this commit, we have introduced a set of assertions to enhance the reliability and correctness of the codebase. Assertions have been strategically placed at critical points within the code to check for expected conditions and ensure that the program behaves as intended.

This commit enhances the robustness of the code and provides a safety net for detecting and debugging issues during development and testing. Assertions will help identify problems quickly, allowing for more efficient bug fixing and maintenance in the future.

aliciamichellew committed:
Merge pull request #2 from aliciamichellew/branch-A-Assertions

Add assertions

aliciamichellew committed:
Fix code quality

aliciamichellew committed:
Merge pull request #3 from aliciamichellew/branch-A-CodeQuality

Fix code quality

aliciamichellew committed:
Add help command

aliciamichellew committed:
Merge pull request #4 from aliciamichellew/branch-C-Help

Add help command

aliciamichellew committed:
Make dialogBox expansible

aliciamichellew committed:
Add styling to chatbox

aliciamichellew committed:
Merge pull request #5 from aliciamichellew/branch-A-BetterGui

Create a better gui

aliciamichellew committed:
Add error handling in parser

aliciamichellew committed:
Refactor code

aliciamichellew committed:
Fix help message

aliciamichellew committed:
Merge pull request #6 from aliciamichellew/A-MoreErrorHandling

Add Error Handling and Fix Bugs

aliciamichellew committed:
Add UI image

aliciamichellew committed:
Change chat radius

aliciamichellew committed:
Change UI image

aliciamichellew committed:
Add User Guide

aliciamichellew committed:
Change note style

aliciamichellew committed:
Change note style

aliciamichellew committed:
Change alert style

aliciamichellew committed:
Merge pull request #7 from aliciamichellew/branch-A-UserGuide

Add User Guide

aliciamichellew committed:
Update README.md

aliciamichellew committed:
Change README.md in the main page

aliciamichellew committed:
Update README.md

aliciamichellew committed:
Update README.md

aliciamichellew committed:
Update README.md

aliciamichellew committed:
Update README.md

aliciamichellew committed:
Fix formatting issue in markdown

aliciamichellew committed:
Fix formatting issue in readme

aliciamichellew committed:
Fix formatting issue in readme

aliciamichellew committed:
Fix error messages

aliciamichellew committed:
Fix parser for deadline and event

aliciamichellew committed:
Remove debug lines

aliciamichellew committed:
Update table of contents
